### PR TITLE
[#44] String[] ElementCollection support

### PR DESCRIPTION
**Pull request by blafond**

### CODE REVIEW

**Comment by gavinking** [comment]
Oh wow! This is a lot of work. It's going to take me some time to review it though!

**Comment by gavinking** [comment]
Out of curiosity: why did you limit it to `String[]`? Aren't other array element types the same?

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/impl/ReactiveCollectionRecreateAction.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+	}
+
+	@Override
+	public AfterTransactionCompletionProcess getAfterTransactionCompletionProcess() {
+		return null;
+	}
+
+	@Override
+	public BeforeTransactionCompletionProcess getBeforeTransactionCompletionProcess() {
+		return null;
```
Comment (DavideD) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/impl/ReactiveCollectionRecreateAction.java:75]: Are these methods correctly implemented?

Comment (DavideD) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/impl/ReactiveCollectionRecreateAction.java:75]: A bit of comments here would help

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/EagerElementCollectionTest.java [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+		private List<String> phones = new ArrayList<>();
+
+		/*
+		 TODO: We also need to check that all the tests pass with the @OrderColumn annotation
+				@ElementCollection(fetch = FetchType.EAGER)
+				@OrderColumn(name = "phone_order")
+				private List<String> workNumbers = new ArrayList<>();
+
+				I don't know if it's better to keep it in this class or a separate one. Your choice.
+		*/
```
Comment (DavideD) [inline hibernate-reactive-core/src/test/java/org/hibernate/reactive/EagerElementCollectionTest.java:239]: I was expecting a test for these cases as well

**Review by DavideD — CHANGES_REQUESTED** [review]
It passes the tests and overall looks fine. That's great!

it seems you left a lot of unnecessary comments and it would be nice to have a couple of additional tests.

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/impl/ReactiveCollectionRecreateAction.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+	}
+
+	@Override
+	public void execute() throws HibernateException {
+		// Unsupported in reactive see reactiveExecute()
+		throw new UnsupportedOperationException( "Use reactiveExecute() instead" );
+	}
+
+	private void preRecreate() {
+		final EventListenerGroup<PreCollectionRecreateEventListener> listenerGroup = listenerGroup( EventType.PRE_COLLECTION_RECREATE );
```
Comment (DavideD) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/impl/ReactiveCollectionRecreateAction.java:90]: We should test these events as well, maybe as a separate issue

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/impl/ReactiveCollectionRecreateAction.java (hunk 1 above)
Comment (blafond) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/impl/ReactiveCollectionRecreateAction.java:75]: I think the overrides in the 3 Actions were redundant... left over from working from the CollectionAction.  Cleaning them up.

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveCollectionPersister.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+ * <p>
+ *     The reactive classes already extend something so we cannot have an abstract class like in ORM.
+ * </p>
+ */
+public interface ReactiveCollectionPersister {
+	CompletionStage<Void> recreateReactive(PersistentCollection collection, Serializable id, SharedSessionContractImplementor session);
+	CompletionStage<Integer> removeReactive(Serializable id, SharedSessionContractImplementor session);
+	CompletionStage<Integer> reactiveDeleteRows(PersistentCollection collection, Serializable id, SharedSessionContractImplementor session);
+	CompletionStage<Integer> reactiveInsertRows(PersistentCollection collection, Serializable id, SharedSessionContractImplementor session);
+	CompletionStage<Integer> reactiveUpdateRows(PersistentCollection collection, Serializable id, SharedSessionContractImplementor session);
```
[the remark this answers is not in the document]
Comment (blafond) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveCollectionPersister.java:25]: Fixed in next commit

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveBasicCollectionPersister.java [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+			CollectionDataAccess cacheAccessStrategy,
+			PersisterCreationContext creationContext) throws MappingException, CacheException {
+		super( collectionBinding, cacheAccessStrategy, creationContext );
+	}
+
+	private ReactiveConnection getReactiveConnection(SharedSessionContractImplementor session) {
+		return ( (ReactiveConnectionSupplier) session ).getReactiveConnection();
+	}
+
+	// See ReactiveOneToManyPersister and superclass to see which method to override and how
```
[the remark this answers is not in the document]
Comment (blafond) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/collection/impl/ReactiveBasicCollectionPersister.java:49]: Removed

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/impl/ReactiveCollectionRecreateAction.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+import org.hibernate.event.spi.PreCollectionRecreateEvent;
+import org.hibernate.event.spi.PreCollectionRecreateEventListener;
+import org.hibernate.persister.collection.CollectionPersister;
+import org.hibernate.reactive.engine.ReactiveExecutable;
+import org.hibernate.reactive.persister.collection.impl.ReactiveCollectionPersister;
+
+
+
+//FIXME: Not sure if we should extend CollectionAction. I guess we can leave it for now
+// and see later if we need to change the methods in the superclass
```
[the remark this answers is not in the document]
Comment (blafond) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/impl/ReactiveCollectionRecreateAction.java:30]: removed

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/ReactiveActionQueue.java [hunk 7]
```diff
@@ -416,29 +416,29 @@ public void addAction(ReactiveEntityUpdateAction action) {
 	 *
 	 * @param action The action representing the (re)creation of a collection
 	 */
-	public void addAction(CollectionRecreateAction action) {
-//		addAction( CollectionRecreateAction.class, action );
-		throw new UnsupportedOperationException();
+	public void addAction(ReactiveCollectionRecreateAction action) {
+		addAction( ReactiveCollectionRecreateAction.class, action );
+//		throw new UnsupportedOperationException();
```
[the remark this answers is not in the document]
Comment (blafond) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/ReactiveActionQueue.java:421]: removed

**Comment by blafond** [comment]
Closing this PR and resubmitting in a bit with better commits/cleaned up code based on comments

**Pull request closed without merging**